core/src/main/scala/kafka/log/stream/s3/metadata/StreamMetadataManager.java

```diff
@@ -74,118 +65,142 @@ public String name() {
-                if (objectMetadata == null) {
-                    // should not happen
-                    LOGGER.error(
-                            "[FetchObjects]: stream: {}, startOffset: {}, endOffset: {}, limit: {}, and search in metadataCache failed with empty result",
-                            streamId, startOffset, endOffset, limit);
-                    throw new IllegalStateException("cannt find object metadata for object: " + object.objectId());
-                }
-                object.setObjectSize(objectMetadata.getObjectSize());
-                object.setCommittedTimestamp(objectMetadata.getCommittedTimeInMs());
-
-            });
-            if (LOGGER.isTraceEnabled()) {
-                LOGGER.trace(
-                        "[FetchObjects]: stream: {}, startOffset: {}, endOffset: {}, limit: {}, and search in metadataCache success with result: {}",
-                        streamId, startOffset, endOffset, limit, rst);
-            }
-            return rst;
-        });
+        return exec(() -> fetch0(streamId, startOffset, endOffset, limit), LOGGER, "fetchObjects")
+                .thenApply(rst -> {
+                    final InRangeObjects inRangeObjects = rst.getKey();
+                    final S3StreamImagePair image = rst.getValue();
+
+                    inRangeObjects.objects().forEach(object -> {
+                        S3Object objectMetadata = image.objectsMetadata().getObjectMetadata(object.objectId());
+                        if (objectMetadata == null) {
+                            // should not happen
+                            LOGGER.error("[FetchObjects]: stream: {}, startOffset: {}, endOffset: {}, limit: {}, " +
+                                            "and search in metadataCache failed with empty result",
+                                    streamId, startOffset, endOffset, limit);
+                            throw new IllegalStateException("can't find object metadata for object: " + object.objectId());
+                        }
+                        object.setObjectSize(objectMetadata.getObjectSize());
+                        object.setCommittedTimestamp(objectMetadata.getCommittedTimeInMs());
+
+                    });
+
+                    if (LOGGER.isTraceEnabled()) {
+                        LOGGER.trace("[FetchObjects]: stream: {}, startOffset: {}, endOffset: {}, limit: {}, " +
+                                        "and search in metadataCache success with result: {}",
+                                streamId, startOffset, endOffset, limit, rst);
+                    }
+
+                    return inRangeObjects;
+                });
     }
 
-    private synchronized CompletableFuture<InRangeObjects> fetch0(long streamId, long startOffset, long endOffset, int limit) {
+    private CompletableFuture<Map.Entry<InRangeObjects, S3StreamImagePair>> fetch0(long streamId, long startOffset, long endOffset, int limit) {
```

Comment:
We can move the logic of filling S3Object information from #fetch to #fetch0, so we don't need to return image.

Reply:
I think the image should be in the same version when get metadata. this pr won't block image update and image retrieve logic. so when the metadata update the images are not the same version which will cause metadata inconsistency.

Reply:
yeah i understand what you said. i will check it out

Reply:
updated